lib/handler.js

```diff
@@ -96,7 +98,23 @@ internals.handler = function (request, callback) {
 
     // Execute handler
 
-    request.route.settings.handler.call(bind, request, reply);
+    const handlerResult = request.route.settings.handler.call(bind, request, reply);
+
+    // Give hapi a fighting chance to deal with uncaught rejections.
+    if (Promises.isThennable(handlerResult)) {
+
+        handlerResult.then(null, (error) => {
+
+            //  Unhandled rejections are always Boom-ified, just like uncaught exceptions.
+            if (error instanceof Error) {
```

Comment:
You certainly aren't supposed to, but out in the wild it happens.  As to whether hapi should deal with the rejection of a non-error, I can't say.  I would probably treat it the same way hapi deals with a non-error thrown in the handler.